### PR TITLE
LPS-186928

### PR DESCRIPTION
**Pull request by marianoalvarosaiz**

### CODE REVIEW

**Comment by catarineaguiar** [comment]
ci:test:echo

**Comment by chaowang-123** [comment]
ci:test:stable

**Comment by chaowang-123** [comment]
ci:help

**Comment by chaowang-123** [comment]
ci:test:security

**Comment by chaowang-123** [comment]
The failure is not related to this pr, @slnn could you help to add QA Passed label?

File: modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/configuration/persistence/listener/LDAPImportConfigurationModelListener.java [hunk 1]
```diff
@@ -40,6 +40,7 @@
  * @author Mariano Álvaro Sáiz
  */
 @Component(
+	enabled = true,
```
Comment (tinatian) [inline modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/configuration/persistence/listener/LDAPImportConfigurationModelListener.java:43]: First, "enable=true" is the default value, no need to set.
Second, please do not have a separate commit to modify a class you just added, please modify it directly in the commit you added.
Third, a component with "enable=true" does not mean it will be activated immediately, you should use immediate=true, I do not understand why it was disabled and then you enable it here.

**Comment by tinatian** [comment]
@marianoalvarosaiz I really do not understand the commit message at https://github.com/liferay-core-infra/liferay-portal/pull/4918/commits/5f48fc3d8c0b50e26083e5e13c14e16930bbea1a.

Before your change, when you modify LDAPImportConfiguration in system settings, you should be able to see the changes in this scheduler job.

The issue you ticket describes is that you are creating LDAPImportConfiguration for each instance, which mean you are having more than one LDAPImportConfiguration, then current code is not working as it did not detect the changes.

The root case is LDAPImportConfiguration is defined as non-factory configuration, but used as a factory configuration which causes you are having more than one LDAPImportConfiguration at the same time.

Fixing in scheduler is just a workaround, you should also fix the original design of LDAPImportConfiguration later. Please also include this information into the commit message.

**Comment by marianoalvarosaiz** [comment]
@tinatian what I mean with:

> To leave behavior as previously, where job is registered based on the system configuration, listen for configuration changes and only reschedule the job if system configuration has changed.

Is that [before](https://github.com/liferay/liferay-portal/blob/7.3.x/modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/configuration/persistence/listener/UserImportConfigurationModelListener.java) we were implementing a ConfigurationModelListener with a listener method that checked if the configuration modified was the one from companyId 0 and only if so the schedulerEngineHelper.register was called. So I'm trying to leave the functionality as it was before, listening for every configuration changed, but only updating the registration if the system company is modified.

I'm not entering into considerations about the design of LDAPImportConfiguration. I agree with you, but I think it should be first fixed to avoid all the errors when having more than one instance and then it can be thought to be redesigned.

> Before your change, when you modify LDAPImportConfiguration in system settings, you should be able to see the changes in this scheduler job.

This is only correct if there are no previous instance settings configurations.

**Comment by marianoalvarosaiz** [comment]
@tinatian I can send a new pull switching from enabled to immediate (my mistake) and adding your comment to the commit comment.

**Comment by tinatian** [comment]
@marianoalvarosaiz 

> I can send a new pull switching from enabled to immediate (my mistake) and adding your comment to the commit comment.
Please do not send a pull with "immediate=true", this is what we avoid using right now. Please see https://liferay.atlassian.net/browse/LPS-180841 and its sub tasks to find a replacement for "immediate=true".


> [before](https://github.com/liferay/liferay-portal/blob/7.3.x/modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/configuration/persistence/listener/UserImportConfigurationModelListener.java) we were implementing a ConfigurationModelListener with a listener method that checked if the configuration modified was the one from companyId 0 and only if so the schedulerEngineHelper.register was called.
This part of logic was already removed by https://liferay.atlassian.net/browse/LPS-175209, if you think it is wrong to do that, you should mention it why it is wrong.

> I'm not entering into considerations about the design of LDAPImportConfiguration. I agree with you, but I think it should be first fixed to avoid all the errors when having more than one instance and then it can be thought to be redesigned.
I am not sure how urgent this is, it seems not that urgent. 
To me, the best way is always to fix things in the root, using a non-factory configuration as a factory configuration is wrong, because of this wrong design, the corresponding scheduler job is not working, scheduler job part is only one of the configuration's usage, you should fix the wrong framework, not to hide the wrong framework by modifying the usage, right?

Please resend.

Thanks.
Tina.

**Comment by marianoalvarosaiz** [comment]
@tinatian https://liferay.atlassian.net/browse/LPS-175209 from a functionality point of view is incorrect. If you define an interval of 10 for the first instance and afterwards an interval of 60 for the second instance the import would be performed for the second instance after 60 minutes. You're somehow recreating the regression that was fixed in https://liferay.atlassian.net/browse/LPS-98420

> I am not sure how urgent this is, it seems not that urgent.
Right now we have failed the SLA of a couple of customers, and another one is on the way.

**Comment by tinatian** [comment]
@marianoalvarosaiz 
I am confused the expected behavior you want.

LPS-98420 only mentioned system settings which LPS-175209 has same behavior. Technically LPS-175209 is a simplify version of LPS-98420. The difference is LPS-98420 can cover the wrong usage of non-factory configuration LDAPImportConfiguration, while LPS-175209 can not. 

While, you just mentioned "If you define an interval of 10 for the first instance and afterwards an interval of 60 for the second instance the import would be performed for the second instance after 60 minutes. ", this is not LPS-98420 targeted at and fix of LPS-98420 shares same issue. This should not be seen as a regression caused by LPS-175209. 

Also having job run based on company's specific configuration is not something you can do based on the implementation of the user import job logic. The job will just iterate all companies and import users. It is in nature a system wide job, not something you should be controlled by a company's specific configuration. 

We do have similar issues like this, the job logic is system wide while we have the configuration is company scope. To fix this in the root, we have to change the job to target at only one specific company. We have similar issues with other jobs, currently the hidden rule is you should not modify the interval on instance settings, the interval will only take effect if you modify on system settings.

To me, the direction to fix this issue is very clear, that is to fix the wrong usage of non-factory LDAPImportConfiguration. Anything you modified in scheduler side is just to cover this wrong usage, not a real fix to the root cause.

Tina.

**Comment by marianoalvarosaiz** [comment]
@tinatian ok, I can try to follow your approach. But, regarding

> It is in nature a system wide job, not something you should be controlled by a company's specific configuration.

how does this match with LXC? In LXC each instance (company) is used to define an environment (uat, production...), so it makes sense to have the ability to control the interval based on a specific company (instance)?

**Comment by tinatian** [comment]
@marianoalvarosaiz Yes, I know it makes sense to control the interval based on specific company, like I said, this is a general issue, we have other similar cases, and I have already mentioned this to Alberto, the team leader for db partition which LXC depends on. This should be considered as a refactor story task for db partition support as it means to rewrite current scheduler jobs to make it support db partition, but until now, I did not know there is a plan for this refactor, I will bring up this again to Alberto.

Tina
